### ISSUE TITLE
JavaScript code evaluation

### ISSUE THREAD

**Issue by Maximus1285** [Huachao/vscode-restclient#521]
It'd be nice to have the option of including JS code snippets to be resolved when sending the request

**Comment by davobutt** [comment]
Yeah this would be really good. Postman's ability to run scripts pre-request is handy for doing simple things like setting dates in the future. I guess there is a slightly elaborate workaround for this in REST client. You could run a local server which executes your script and responds in json which the REST client file can extract the variables from?

**Comment by jnorkus** [comment]
I am about to migrate from Postman. This is the only feature missing. I don't think running a local server would convince my team.

**Comment by xslim** [comment]
I need to apply encription / decryption for all requests, so pre&post JS processing would be awesome!

**Comment by RimiX2** [comment]
It would be POSTMAN-killer feature!

**Comment by robermorales** [comment]
Yeah I would also like to have some scripting abilities.

**Comment by robermorales** [comment]
At least, some more complex variable processing to cover most of cases: like replacing only a substring of a variable, things like that. Some "templating" abilities can be enough.

**Comment by NickClark** [comment]
It would be handy, for JSON, to have a [`jq`](https://stedolan.github.io/jq/) style post filter ability.

**Comment by Maslow** [comment]
I am looking for this feature then arrived here.  It would be handy.

**Comment by Maslow** [comment]
I am looking for this feature then arrived here. It would be handy.

Add custom-filter or hooks during req & res duration.

**Comment by afeblot** [comment]
> It would be handy, for JSON, to have a [`jq`](https://stedolan.github.io/jq/) style post filter ability.
yes, it would!! Ideally, integrated in the rest client output pane. Or a way to integrate or "pipe" to existing vscode jq extensions such as jq-vscode or VSCode js playground

**Comment by showenx** [comment]
First of all, love this extension, I was searching some solution for code blocks in the extension, I just want to throw some ideas, maybe helpful or not.

Is it possible for variable resolution, using IIFE style ( similar to other ****Monkey chrome extensions) to wrap up user's custom code block?

e.g.
```javascript
(function (param) {
    'use strict';

    /*
     * user custom code block goes here
     */

    return param + '->output';
})('input');
```

**Comment by asos-gurpreetsingh** [comment]
on the home page for this extension it says "Generate code snippets for HTTP request in languages like Python, JavaScript and more" but have not seen any examples or documentation around this feature

**Comment by davobutt** [comment]
> on the home page for this extension it says "Generate code snippets for HTTP request in languages like Python, JavaScript and more" but have not seen any examples or documentation around this feature

That is a feature that turns your request in to code that sends the same request.. If you do the "Rest Client: Generate code snippet" action you'll see what it does.

**Comment by kibiz0r** [comment]
I had an idea on this... What if we treated code blocks similarly to request blocks?

You give them a name, they are executed linearly top-to-bottom, with the stdout cached -- and if it can be parsed as JSON then you can treat it as JSON when referencing it later on.

We could even have some double-meaning with the language annotation:
1. It gives us nice Markdown syntax highlighting
2. It tells us what interpreter to use, effectively a #! shebang

Example `foo.http`:

````

# @name randomData
```ruby
mac = 6.times.map { ("%02x" % rand(256)).upcase }.join(":")
name = ["dingus", "dorkus", "derp"].sample
puts({ mac: mac, name: name }.to_json)
```

###

POST localhost:3000/devices
content-type: application/json

{
  "name": {{randomData.name}},
  "macAddress": {{randomData.mac}}
}

````

**Comment by piyushchauhan2011** [comment]
Any updates on this ? like running pre and post js code

**Comment by asampal** [comment]
@piyushchauhan2011, you can take a look at httpyac. I switched to that quite a while ago. It's more actively maintained and has more features. https://github.com/AnWeber/httpyac

**Comment by piyushchauhan2011** [comment]
Thanks for the links @asampal . Hope restclient also add similar features, but httpyac seems to have good progress at present moment

**Comment by matejsp** [comment]
I would be very nice to have calls compatible with intelij http client:
https://www.jetbrains.com/help/idea/javascript-api-supported-by-http-client.html#exec-methods

```
< {%
  before javascript
}
POST http://localhost:8001/echo
Content-Type: application/json

{
  "message": "Hello"
}

> {%
  after javascript
}
```
